### PR TITLE
Support relational operators for decimal type

### PR DESCRIPTION
**Pull request by nartal1**
Support relational operators for decimal type. Added integration tests to verify corner cases.

### CODE REVIEW

File: integration_tests/src/main/python/data_gen.py [hunk 1]
```diff
@@ -208,6 +209,15 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
...
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, nullable=True):
+        super().__init__(DecimalType(7,3), nullable=nullable)
+
+    def start(self, rand):
+        self._start(rand, lambda : Decimal(str(round(random.uniform(2000, 4000), 2))))
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:219]: I really only see one good choice when generating random data for Decimal.  We need to use `sre_yield` like with the StringGen.  We need to create a pattern that will work.  Essentially we need to have a pattern like

```
"-?[0-9]{1," + str(precision-scale) + "}.[0-9]{0," + str(scale) + "}"
```

Note that by default Spark does not support a negative scale (That is controlled by `park.sql.legacy.allowNegativeScaleOfDecimal`) but we do need to support it for testing.  So the above pattern only works for positive scale.  For negative scale we would need something more like.

```
"-?[0-9]{1,"+str(precision) + "}e" + str(-scale)
```

Note I have not tested these, this is just off the top of my head.

File: integration_tests/src/main/python/data_gen.py [hunk 2]
```diff
@@ -685,6 +697,8 @@ def gen_scalars_for_sql(data_gen, count, seed=0, force_no_nulls=False):
 # a selection of generators that should be orderable (sortable and compareable)
 orderable_gens = [byte_gen, short_gen, int_gen, long_gen, float_gen, double_gen,
         string_gen, boolean_gen, date_gen, timestamp_gen]
+orderable_gens_with_decimal_gen = [byte_gen, short_gen, int_gen, long_gen, float_gen, double_gen,
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:700]: nit:  It is a lot cleaner to do

```
orderable_gens_with_decimal_gen = orderable_gens + [decimal_gen]
```

Ideally I'd like to see a comment that `orderable_gens_with_decimal_gen` would go away once all we can add decimal types to `orderable_gens` directly.

File: integration_tests/src/main/python/data_gen.py [hunk 3]
```diff
@@ -668,8 +678,10 @@ def gen_scalars_for_sql(data_gen, count, seed=0, force_no_nulls=False):
 boolean_gen = BooleanGen()
 date_gen = DateGen()
 timestamp_gen = TimestampGen()
+decimal_gen = DecimalGen()
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:681]: We are going to need more than just one.  We really are going to need a few different sets of decimal_gens. A default set that work with `park.sql.legacy.allowNegativeScaleOfDecimal` disabled, and others that work when it is enabled.

File: integration_tests/src/main/python/cmp_test.py [hunk 4]
```diff
@@ -57,7 +57,9 @@ def test_ne(data_gen):
                 f.col('b') != f.lit(None).cast(data_type),
                 f.col('a') != f.col('b')))
 
-@pytest.mark.parametrize('data_gen', orderable_gens, ids=idfn)
+@incompat
+@pytest.mark.allow_non_gpu('ProjectExec', 'Alias')
```
Comment (revans2) [inline integration_tests/src/main/python/cmp_test.py:61]: These should be changed now.  It is not incompat and Project and Alias are on the GPU in all cases.

Comment (nartal1) [reply integration_tests/src/main/python/cmp_test.py]: Done

File: integration_tests/src/main/python/data_gen.py (hunk 3 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Added for default case, positive and negative scale. Thanks!

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Removed this as all support for all operators which take in `orderable_gens` is done. Have added for `eq_gens` as there is no decimal type support for `EqualNullSafe`

File: integration_tests/src/main/python/data_gen.py [hunk 5]
```diff
@@ -208,6 +209,15 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
     def start(self, rand):
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, nullable=True):
```
[the remark this answers is not in the document]
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Your PR covered this.

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Your PR covered this.

**Comment by nartal1** [comment]
build

File: integration_tests/src/main/python/cmp_test.py [hunk 6]
```diff
@@ -31,7 +31,7 @@ def test_eq(data_gen):
                 s2 == f.col('b'),
                 f.lit(None).cast(data_type) == f.col('a'),
                 f.col('b') == f.lit(None).cast(data_type),
-                f.col('a') == f.col('b')))
+                f.col('a') == f.col('b')), conf={'spark.sql.legacy.allowNegativeScaleOfDecimal': 'true'})
```
Comment (revans2) [inline integration_tests/src/main/python/cmp_test.py:34]: If this is going to be common for a lot of decimal tests can we make it something in data_gen that can be used by multiple tests?

File: integration_tests/src/main/python/data_gen.py [hunk 7]
```diff
@@ -708,13 +745,16 @@ def gen_scalars_for_sql(data_gen, count, seed=0, force_no_nulls=False):
 # TODO add in some array generators to this once that is supported for sorting
 # a selection of generators that should be orderable (sortable and compareable)
 orderable_gens = [byte_gen, short_gen, int_gen, long_gen, float_gen, double_gen,
-        string_gen, boolean_gen, date_gen, timestamp_gen, null_gen]
+        string_gen, boolean_gen, date_gen, timestamp_gen, null_gen] + decimal_gens
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:748]: Do the sort tests work with this change?

File: integration_tests/src/main/python/data_gen.py [hunk 8]
```diff
@@ -208,6 +209,37 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
     def start(self, rand):
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+
+#Maximum number of decimal digits a Long can represent is 18.
+DECIMAL_MIN = -(1 << 59)
+DECIMAL_MAX = (1 << 59)  #approximate
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:215]: I don't think these work the way we want them to because the maximum and minimum values are dependent on the precision and scale.  These values are invalid for some combinations of these.

I would really prefer to have us generate a string that represents the correct values, using something like the pattern below.

```
Decimal(('9'*precision) + 'e' + str(-scale)), Decimal('-' + ('9' * precision) + 'e' + str(-scale))
```

File: integration_tests/src/main/python/data_gen.py [hunk 9]
```diff
@@ -208,6 +209,37 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
...
+    def __init__(self, precision=None, scale=None, nullable=True, special_cases=[]):
+        if precision is None:
+            precision = 18
+            scale = 0
+            special_cases = [Decimal('-1'), Decimal('0'), Decimal('1'), Decimal(DECIMAL_MIN), Decimal(DECIMAL_MAX)]
+        super().__init__(DecimalType(precision, scale), nullable=nullable, special_cases=special_cases)
+        self._scale = scale
+        self._precision = precision
+        if (scale > 0):
+            pattern = "[0-9]{1,"+ str(precision - scale) + "}\.[0-9]{0," + str(scale) + "}"
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:227]: I made a mistake here.  If scale > precision, then this pattern is not going to work.  I think we might just want to use the other pattern all of the time.

File: integration_tests/src/main/python/data_gen.py [hunk 10]
```diff
@@ -208,6 +209,37 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
...
+#Maximum number of decimal digits a Long can represent is 18.
+DECIMAL_MIN = -(1 << 59)
+DECIMAL_MAX = (1 << 59)  #approximate
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, precision=None, scale=None, nullable=True, special_cases=[]):
+        if precision is None:
+            precision = 18
+            scale = 0
+            special_cases = [Decimal('-1'), Decimal('0'), Decimal('1'), Decimal(DECIMAL_MIN), Decimal(DECIMAL_MAX)]
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:222]: Even -1 and 1 do not work all of the time.  a precision of 1 and scale of -3 would be an example where this would not work. It can only hold multiples of 100 up to 900.

Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Sorry, I don't understand above statement.  precision of 1 and scale of -3 is `1000`  and I am not getting why it can hold multiples of 100. Could you please explain. I tried setting precision and scale as mentioned above and it passed. So I am missing something.

File: integration_tests/src/main/python/data_gen.py (hunk 9 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: done

File: integration_tests/src/main/python/data_gen.py (hunk 8 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: done

File: integration_tests/src/main/python/data_gen.py (hunk 7 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py:744]: It doesn't affect sort tests as they don't use `orderable_gens` instead use `orderable_gen_classes`

File: integration_tests/src/main/python/cmp_test.py (hunk 6 above)
Comment (nartal1) [reply integration_tests/src/main/python/cmp_test.py]: Done

File: integration_tests/src/main/python/data_gen.py (hunk 7 above)
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py:744]: It doesn't affect sort tests as they don't use `orderable_gens` instead use `orderable_gen_classes`

File: integration_tests/src/main/python/data_gen.py (hunk 10 above)
Comment (revans2) [reply integration_tests/src/main/python/data_gen.py]:
```
>>> df = spark.createDataFrame(data=[(Decimal('-1')), (Decimal('0'))], schema=DecimalType(7, -3))
>>> df.show()
+-----+
|value|
+-----+
| 0E+3|
| 0E+3|
+-----+
```

The details of -2 vs -3 scale don't matter, what matters is -1 becomes 0 for any scale that is negative.  For a Decimal where the precision is larger than the scale this actually becomes an error'

```
>>> df = spark.createDataFrame(data=[(Decimal('-1')), (Decimal('0'))], schema=DecimalType(7, 7))
>>> df.show()
20/11/25 12:52:57 ERROR Executor: Exception in task 0.0 in stage 32.0 (TID 125)
java.lang.ArithmeticException: Decimal precision 8 exceeds max precision 7
	at org.apache.spark.sql.types.Decimal.set(Decimal.scala:122)
	at org.apache.spark.sql.types.Decimal$.apply(Decimal.scala:571)
```

We cannot blindly add in -1 as a special case.

Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py]: Thanks for the exmaples. I have removed `-1` and `1` from special cases. Added another generator where scale and precision are same.

File: integration_tests/src/main/python/data_gen.py [hunk 11]
```diff
@@ -208,6 +209,33 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
     def start(self, rand):
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, precision=None, scale=None, nullable=True, special_cases=[]):
+        if precision is None:
+            #Maximum number of decimal digits a Long can represent is 18
+            precision = 18
```
[the remark this answers is not in the document]
Comment (nartal1) [reply integration_tests/src/main/python/data_gen.py:217]: Agreed. Selecting correct precision and scale for each test would be good.

**Comment by sperlingxx** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
@revans2 I think I have addressed remaining review comments. Please take another look.

**Review by revans2 — APPROVED** [review]
